### ISSUE TITLE
C# Documentation Comments

### ISSUE THREAD

**Issue by Mantissa-23** [godotengine/godot-proposals#8269]
### Describe the project you are working on

Currently working on a movement shooter, but this applies to _any_ project using C#.

### Describe the problem or limitation you are having in your project

C# docstrings do not apply to inspector property documentation-on-hover, unlike [GDScript documentation comments](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_documentation_comments.html).

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Seeing docstrings in the inspector is valuable for both solo devs, and particularly for teams, where nontechnical team members often need to interact with classes generated by programmers, and the name of a property alone may not fully describe its functionality.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

While I'm unclear on the implementation details, the end-goal is fairly straightforward- Bring C# documentation comments to feature parity with GDScript documentation comments.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Not that I'm aware of.

### Is there a reason why this should be core and not an add-on in the asset library?

For the same reason that both C# and GDScript documentation comments are core.

**Comment by ttencate** [comment]
I went looking for the same just now, and was surprised to find that (a) this functionality doesn't exist (or is broken in 4.2-beta3?), (b) there's no issue about it, and (c) there wasn't even a proposal until 5 hours ago (unless my search-fu failed me). So big thumbs up from me.

**Comment by Calinou** [comment]
> I went looking for the same just now, and was surprised to find that (a) this functionality doesn't exist (or is broken in 4.2-beta3?), (b) there's no issue about it, and (c) there wasn't even a proposal until 5 hours ago (unless my search-fu failed me). So big thumbs up from me.

This is already an open discussion: https://github.com/godotengine/godot-proposals/discussions/5675

**Comment by ttencate** [comment]
@Calinou Given the number of upvotes on that discussion, would it make sense to convert it (back) to an issue and close this one as a duplicate?

**Comment by Calinou** [comment]
> @Calinou Given the number of upvotes on that discussion, would it make sense to convert it (back) to an issue and close this one as a duplicate?

Discussions can't be converted to issues (only the other way around is possible). You can create a new issue from a discussion, but this will just quote the discussion and won't create a redirect or transfer votes/reactions.

**Comment by ttencate** [comment]
Alright, let me take a stab at detailing the proposal.

Besides inspector tooltips (which would indeed be a major use case), GDScript doc comments also trigger generation of in-editor help pages. There are probably more places where doc comments are, or could be, used, e.g. in autocomplete tooltips. Since the rename of the issue, that's also in scope for this proposal, but it will just come along for free.

### Documentation comment syntax

Reference:

* [Documentation comments in C#](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/documentation-comments)

* [Documentation comments in GDScript](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_documentation_comments.html)  
  It's worth noting that the documentation for script _members_ doesn't support the distinction between summary, description and tutorial links. It's just all concatenated into a single string, which may contain BBCode.

Of course, we wouldn't want to implement a parser for C# doc comments – Microsoft already did that. We should simply piggy-back on that, and convert the XML to BBCode that the editor understands. As a first attempt, just passing all XML through verbatim would already support doc comments that are just plain text, although many IDEs automatically insert the (required?) `<summary></summary>` tags so those would look messy.

Fortunately the [list of common tags](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/documentation-comments#d31-general) is fairly small. The following translations could be applied:

| XML | BBCode | Notes |
|-----|--------|-------|
| `<c>...</c>` | `[code]...[/code]` | |
| `<code>...</code>` | `[codeblock]...[/codeblock]` | |
| `<example>...</example>` | `[b]Example:[/b][br]...` | |
| `<exception cref="NAME">...</exception>` | `[b]Throws [NAME]:[/b] ...` | Godot doesn't have proper exception handling, but showing explicitly documented exceptions in the docs would be good. |
| `<include/>` | | Not supported (include verbatim) |
| `<list>` | | Not supported (include verbatim) |
| `<para>...</para>` | `...[br]` | |
| `<param name="NAME">...</param>` | `[param NAME]: ...[br]` | |
| `<paramref name="NAME"/> | `[param NAME]` | |
| `<permission/>` | | Not supported (include verbatim) | |
| `<remarks>...</remarks>` | | Content appended (after `[br]`) to description |
| `<returns>...</returns>` | `[b]Returns:[/b] ...` | |
| `<see cref="MEMBER" href="URL" langword="KEYWORD"/>` | `[member CURRENT_CLASS.MEMBER]`  | `href` and `langword` not supported |
| `<seealso cref="MEMBER" href="URL"/>` | `See [member CURRENT_CLASS.MEMBER][b]`  | `href` not supported |
| `<summary>...</summary>` | | For classes, appended to `brief_description`, otherwise to `description` | `<typeparam name="NAME">...</typeparam>` | `[code]NAME[/code]: ...[br]` | Not supported in Godot but still belongs in the documentation |
| `<typeparamref name="NAME"/>` | `[code]NAME[/code]` | Not supported in Godot but still belongs in the documentation |
| `<value>...</value>` | `[b]Value[/b]: ...[br]` | |

### Implementation

Internally, documentation data is stored in the `DocData` class (technically a class but used only as a namespace) in [`core/doc_data.h`](https://github.com/godotengine/godot/blob/master/core/doc_data.h). The inspector uses this to create its tooltips, and the in-editor help uses it too. The entry point is `DocData::ClassDoc`, which contains vectors of the documentation of its members.

Both the inspector and the in-editor help obtain this `ClassDoc` (or rather, a vector of them, because inner classes) by calling the virtual function [`Script::get_documentation()`](https://github.com/godotengine/godot/blob/09946f79bd8215b2c6332de8821737580909a91c/core/object/script_language.h#L143), so the main task would be to implement this for `CSharpScript`, [aptly marked `TODO`](https://github.com/godotengine/godot/blob/09946f79bd8215b2c6332de8821737580909a91c/modules/mono/csharp_script.h#L163-L167).

In GDScript, the documentation is stored in a private member variable `docs`, which is populated through [`_add_doc`](https://github.com/godotengine/godot/blob/09946f79bd8215b2c6332de8821737580909a91c/modules/gdscript/gdscript.h#L157), which is called from [`GDScriptDocGen::generate_docs`](https://github.com/godotengine/godot/blob/09946f79bd8215b2c6332de8821737580909a91c/modules/gdscript/editor/gdscript_docgen.cpp). It receives the parse tree from the GDScript parser, which already parsed the doc comments and attached them to the parse tree. This is where the C# implementation would diverge.

By setting [`GenerateDocumentationFile`](https://learn.microsoft.com/en-us/dotnet/core/project-sdk/msbuild-props#generatedocumentationfile) in the `.csproj` file, the compiler will emit an XML file containing all the documentation for the assembly. This ends up in `.godot/mono/temp/bin/Debug/PROJECTNAME.xml` alongside the `PROJECTNAME.dll`. (By default, the compiler also spits out a warning for every undocumented public field. That should be disabled or made opt-in.)

After successfully (re)loading the assembly, the mono module would attempt to load and parse an XML file next to the DLL with the same name, apply the above transformations, and add the documentation to each class that it knows about. This code would of course be guarded by `#ifdef TOOLS_ENABLED` because the documentation only exists in editor builds.

**Comment by Mantissa-23** [comment]
Whoops, didn't realize that existed. I swore I dug pretty hard for this. Feel free to close this issue, or reuse it for the implementation of the proposal.

**Comment by ttencate** [comment]
> Do we expect the BBCode syntax to be 100% identical to the one used in GDScript, including for references to other properties (`[member something]`) or functions (`[method something]`)?

Using BBCode in the C# doc comments directly didn't even occur to me. It would make implementation simpler, but it would
* feel weird for C# developers,
* not have IDE suppport,
* not play nice if the doc comments are also to be used in the .NET ecosystem.
So I don't think this is the right path.

**Comment by geekley** [comment]
> Using BBCode in the C# doc comments directly didn't even occur to me.

I expect that BBCode inside C# doc comments is undesired and ***must not*** be supported at all. Any part with characters that BBCode would interpret (e.g. `[`, `]`) ***must*** be properly escaped. This is doubly important inside `<c>` and `<code>` tags, but also everywhere else.

It's also best to define how any other tags not listed are handled. Should they just be removed along with contents? Or included verbatim, maybe as code? e.g: `[code]<my-tag>[/code]...[code]</my-tag>[/code]`

Gotta also consider how "open documentation" (in Godot Script docs view) command (e.g. when right-clicking an exported variable) should likely support C# docs too after this, and it may have clickable links to other types/fields/etc.
Likely there will need to be some conversions (e.g. PascalCase to underscore_case and maybe other equivalences) when referring to engine types/fields/etc. I don't mean in the displayed text, just in links - so they work properly.

Also, it is possible to support `href` and `langword` through URLs. URLs do work in GDScript if you use BBCode syntax like `[url=https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_basics.html#keywords]self[/url]`.

**Comment by Mantissa-23** [comment]
> I expect that BBCode inside C# doc comments is undesired and _**must not**_ be supported at all.

Is there a good reason for this other than "because it would be mixing two different markup syntaxes"? The only reason I can think of this being an issue is if we were to parse non-user generated docstrings, i.e. those in libraries, which I have to imagine we'd want to avoid at all costs for performance and conciseness reasons.

As far as I'm aware, C#'s official documentation doesn't mention any of the patterns in BBCode being used for anything. At first blush they seem compatible?

**Comment by geekley** [comment]
Mixing 2 different markup systems is enough good reason for not doing it. If you really wanted to do it, the implementation would have to be done right (think cases like `<c>[b]...[/b]</c>` and `[code]<see.../>[/code]`; what do they mean?). Doing it right would complicate the implementation and even so it would only cause confusion.
Not to mention the reasons already pointed in the comment above mine.

If you're using xmldoc, you expect xmldoc syntax, ***consistently***. If this proposal wants to support BBCode syntax, then would be better to completely separate it from xmldoc and just implement it with attributes, like:
`[Tooltip("[b]BBCode here[/b] etc")]`
But this would probably not be what people want. They want xmldoc compatibility.

**Comment by geekley** [comment]
> The following translations could be applied: [...] apply the above transformations

Not sure if I misunderstood the proposed solution, but honestly, converting xmldoc to BBCode to reuse GDScript's implementation seems like a good idea at first, but it might end up biting us in the leg later. If you're gonna take the effort to properly handle all cases and escapes correctly, then it's better to instead use this effort to just do it properly and parse the XML directly into whatever is used to "low-level" render the formatted docs.

Even if it's inevitable to use BBCode under the hood after parsing the XML, ***C# docs implementation should have no relation to GDScript docs implementation***. I mean it shouldn't generate things like `[method something]`, instead both C# and GDScript (and theoretically any other language) should be parsed into whatever links and formatting are needed to give it a similar look and behavior, in a language-agnostic way. Even if a `ClassDoc` ends up having to use vanilla BBCode under the hood for rendering (I have no idea if that's the case) - you should not need to parse any BBCode at all to create a `ClassDoc`. That would be just asking for trouble and bugs.

Again, sorry if misunderstood the proposed solution, just thought this might need clarification.

**Comment by ttencate** [comment]
Good point. I tend to agree; however I'm not sure this can be implemented in a backwards compatible way. If any of the docs machinery is accessible from outside the engine, like from GDScript or GDExtension, it may get difficult to refactor it to make it markup language agnostic. Would need to check.

**Comment by ttencate** [comment]
@nanodeath It would be for all doc comments, on classes, methods and fields (including exported fields).

**Comment by 31** [comment]
Here's a PR in progress to add inspector doc tooltips based on C# docs. Mentioning it here because there were no links, and want to make it easier to find in the future: 🙂 

* https://github.com/godotengine/godot/pull/83505

**Comment by geekley** [comment]
Just remembered. It's important that the solution for docs is language-independent, because it needs to be implemented not only in C#, but GDShader as well. The `uniform` fields are shown in the inspector, and they should also allow documentation. I'd expect syntax like this (both triple-slash and `/**` supported):
```glsl
shader_type canvas_item;

/** Color of the outline. */
uniform vec4 color: source_color = vec4(1);

/// Thickness of the outline
uniform float thickness: hint_range(0, 127) = 4;

// etc...
```
Or maybe you could even use the same XMLDoc syntax as in C#, if that's simpler.

**Comment by ttencate** [comment]
That totally makes sense in some situations, but leads to needless duplication in others. Maybe we can have both? If there's only a doc comment, that will be used as the description, but the Export attribute can be used to override it. (Note that the tooltip is currently not separate from the description in the help browser, where developer-focused comments would probably make more sense...)

**Comment by mads-fs** [comment]
> That totally makes sense in some situations, but leads to needless duplication in others. Maybe we can have both? If there's only a doc comment, that will be used as the description, but the Export attribute can be used to override it. (Note that the tooltip is currently not separate from the description in the help browser, where developer-focused comments would probably make more sense...)

If both are desired in that way sure I suppose. I'd personally find it unnecessary but honestly as long as I can separate developer concerns from designer concerns then that's all good by me. Code comments should, in my opinion, pertain to developer concerns, not designer concerns.

But it sounds like we largely agree that it's useful at least.

**Comment by geekley** [comment]
@mads-fs Not sure if I misunderstand the issue, but in your example, the TODO comment would not mess up the tooltip because it's only double-slashed `// TODO`, not triple-slashed `/// TODO`. If tooltips are parsed from the XMLDoc, then it's implied only `///` (and `/** ... */` I think?) are considered, not `//` or `/* ... */`.

Regarding attribute vs XMLDoc, I'll just repeat what I said above:
- BBCode should NOT be used for formatting on XMLDoc. If BBCode is used, it has to be only via attribute. XMLDoc must be formatted with the XML tag, the expected way.

Also agree with @ttencate that the ideal solution might be to have both, as:
- The advantage of XMLDoc is that a single documentation works as both API documentation for programmers and editor documentation for designers. Not duplicating makes things easier. It's also consistent with GDScript's use of `##` documentation comments in this sense, as well as with how built-in core classes' API docs is the same on both editor and code.
- The advantage of tooltip (though I'd argue a different attribute to apply together with Export is likely better) is that you could still have this separation when needed. You may want to describe programmer API differently from designer tooltip help. E.g. in Unity you can do this, because an attribute is used for tooltips; and I've made use of this difference when making a Unity editor plugin, though in practice both docs didn't end up with that big a difference, so maybe it's not a *necessity*.

So, both is ideal, but if I have to pick only one solution, I'd probably prefer the XMLDoc for simplicity and consistency.

**Comment by mads-fs** [comment]
> @mads-fs Not sure if I misunderstand the issue, but in your example, the TODO comment would not mess up the tooltip because it's only double-slashed `// TODO`, not triple-slashed `/// TODO`. If tooltips are parsed from the XMLDoc, then it's implied only `///` (and `/** ... */` I think?) are considered, not `//` or `/* ... */`.

Yes that is misunderstanding. It's about separating the concerns of developers with the concerns of designers. The solution is fine in isolation, however if you work in teams (as I've done professionally) then you want to get non-programmers into the editor as much as possible to do as much as they can without a programmer present. To do so you often also have to approach things like comments, tooling, etc. different.

So to be able to leave code comments for programmers and comments for designers as separate things is very desirable in larger productions. Ultimately my suggestion would allow for both code comments and editor property descriptions without the two taking more space in the code than they need to (UX).

**Comment by geekley** [comment]
I agree that it's likely important to be able to have both docs (specially for plugins and larger teams) ...
But ***only as long as duplication is not required***. Otherwise the extra work would only demotivate people from documenting at all (definitely a bad thing, as people already don't document stuff properly as-is).

So the ideal is indeed using XMLDoc, but allowing to override it if needed.

What I absolutely don't want is ***having*** to do this extra work everywhere, like I had to do in Unity:
```cs
///<summary>
/// Same super-long documentation text twice.
/// Blah blah blah.
///</summary>
[Tooltip(
"Same super-long documentation text twice.\n" +
"Blah blah blah."
)]
[Export] public string someField;
```

**Comment by mads-fs** [comment]
> I agree that it's likely important to be able to have both docs (specially for plugins and larger teams) ... But _**only as long as duplication is not required**_. Otherwise the extra work would only demotivate people from documenting at all (definitely a bad thing, as people already don't document stuff properly as-is).
> 
> So the ideal is indeed using XMLDoc, but allowing to override it if needed.
> 
> What I absolutely don't want is _**having**_ to do this extra work everywhere, like I had to do in Unity:
> 
> ```cs
> ///<summary>
> /// Same super-long documentation text twice.
> /// Blah blah blah.
> ///</summary>
> [Tooltip(
> "Same super-long documentation text twice.\n" +
> "Blah blah blah."
> )]
> [Export] public string someField;
> ```

That's a fair point. Long as both are available in that way then that should work I think.

**Comment by Creta5164** [comment]
Here, as a developer in the .NET ecosystem who has been following the discussion via an alert subscription, I thought I'd comment in case there are any insights that participants are missing.

First of all, let me say that I'm not a big fan of XML syntax (I'm jealous of GDScript's comment documentation and jsDoc), but the metadata that XMLDoc's specification allows to be expressed is quite powerful (dependency tracking, inheritance, etc.).

So if we assume that Godot's specification for documentation annotations is meant to support XMLDoc's specification, the design direction of the two is different enough to make you wonder, "Is this necessary for Godot's core?".

On the flip side, the .NET ecosystem has also grown from this foundation, so I agree with other participants that it is difficult to use BBCode for XMLDoc, and that it should not be used.

For example, the DocFX documentation tool, which is predominantly used in the .NET ecosystem, would not be able to analyze BBCode, which is not a standard specification from a .NET perspective.

So I would ask the participants in this discussion to consider the backgrounds of both Godot and .NET in order to reach a meaningful conclusion.

It's a bit of a stretch in my opinion, but I think it would be great to look at the example of GDExtension, which plugs deeply into Godot, and it would be great to get insights from similar examples for C#, which is a first language to build with, but not GDScript.

> *This is machine translated, may have typo or grammar error.*

**Comment by thetawave** [comment]
100% should parse C# XML documentation. [Hotchocolate](https://github.com/ChilliCream/graphql-platform) is an example library which generates (graphQL) schema documentation from C# XML documentation. It works well in production.

**Comment by Mantissa-23** [comment]
Just wanted to give a thank you to @magian1127 for the PR for this. I just remembered this and it's wild to see someone's already gone and implemented it.

**Comment by Mantissa-23** [comment]
@dante-signal31 I agree it seems stalled, though I'm not privy to what's going on with the PR or why. The contributor's a volunteer, not a paid contributor, they probably had something get in the way IRL. Might be worth offering to pick it up if you have the time.

**Comment by dante-signal31** [comment]
Sadly, the knowledge level needed is still above mine @Mantissa-23 . Nowadays, I can only say thanks to whoever able to implement something like this.

**Comment by sdepouw** [comment]
Chiming in to say that I would very much desire this! For me it feels like quite the roadblock in terms of using C# (a language I'd rather use since I'm super familiar with it).

As for how to implement, throwing my two cents in to say "don't change how C# documentation works literally everywhere else" and just support the XML Documentation Comments. (i.e. `<summary>` tag and friends).

**Comment by stephen-polyxia** [comment]
A good starting point would be to use an attribute to generate a simple text based description as suggested elsewhere in this thread. Something like:

```
[Export]
[Tooltip("This is my description")]
public string MyValue { get; set; }
```

I assume this would be a lot simpler to implement than parsing the XML, which I would guess is the reason it has been 2 years since this issue was opened but we still have no support for documentation on properties.

If the XML version is implemented later, backwards compatibility should be simple enough. Some ideas on how that could be achieved:

- Allow both methods and make XML docs supersede the `[Tooltip]` attribute if both are present.
- Give a compiler error if both `[Tooltip]` attribute and XML comment are present.
- Merge the `[Tooltip]` attribute contents with the XML contents.

**Comment by alexcamerondev** [comment]
Baffled why this isn't already in the engine.

**Comment by geekley** [comment]
> If the XML version is implemented later, backwards compatibility should be simple enough. Some ideas on how that could be achieved:
> - Allow both methods and make XML docs supersede the `[Tooltip]` attribute if both are present.
> - Give a compiler error if both `[Tooltip]` attribute and XML comment are present.
> - Merge the `[Tooltip]` attribute contents with the XML contents.

@stephen-polyxia No, superseeding has to be the opposite way:
- if both are present, XMLDoc controls docs shown in code while Tooltip overrides it for what's shown in the inspector.
- if only XMLDoc is present, it's shown in code and the inspector should parse XML to show docs in the inspector  
  (but I agree this could be implemented later)  
  (if XMLDoc in the inspector is undesired, an empty `[Tooltip("")]` can be used to make it not show a tooltip at all)
- if only Tooltip is present, it's shown in the inspector; code docs will ignore it and that's by design

Compile error and merge are bad solutions. The whole point of having 2 ways is so you allow different docs for the programmer in code and for the designer in the inspector, as [each could be concerned with different aspects of the API](#issuecomment-2159127401).

As for syntax, it's important that comments use XMLDoc syntax, not BBCode. When Tooltip is present, it overrides the docs shown in the inspector; and Tooltip ***could*** use BBCode if that's easier to implement. But the comment docs, if parsed, definitely have to use only XMLDoc syntax so as to not clash with other tools that use that standard for docs.
